[RFC 0109] Nixpkgs Generated Code Policy (#109)
NixOS/rfcs: rfcs/0109-nixpkgs-generated-code-policy.md

---
feature: nixpkgs-generated-code-policy
start-date: 2021-10-12
author: John Ericson (@Ericson2314)
co-authors: (find a buddy later to help out with the RFC)

> shepherd-team: @L-as @sternenseemann @tomberek @DavHau
> shepherd-leader: @sternenseemann

Reviewer: interpreting @KFearsoff's comment as accepting the nomination...

>@lheckemann proposed me as a shepherd, so I could try doing that. Not tomorrow, but after that.

```suggestion
shepherd-team: @L-as @kfearsoff @tomberek @DavHau
shepherd-leader: @kfearsoff
```

Reviewer 2: Then I'd take over for @Ericson2314 at the same time?

Reviewer 3: Woah there, haven't agreed to be the shepherd leader yet. Need to understand the responsibilities first, which I don't understand too well right now.

Reviewer 4: I had the same reaction when I was nominated as shepherd chair for rfc133... and had zero prior involvement in the rfc process.

It really just means being the person who's responsible for scheduling meetings, taking notes, and not dropping the ball (which I did, a few times).

Maybe "shepherd secretary" would be a bit less scary to people getting roped in for the first time.

Reviewer: Understandable, sorry about that! We'll try to take previous experience into account in future shepherd team decisions. Tom has volunteered to take the lead on this one.

```suggestion
shepherd-team: @L-as @kfearsoff @tomberek @DavHau
shepherd-leader: @tomberek
```

Reviewer 2: I'm still unclear on why I'm being removed altogether?

> author: John Ericson (@Ericson2314)
> co-authors: (find a buddy later to help out with the RFC)
> shepherd-team: @L-as @sternenseemann @tomberek @DavHau
> shepherd-leader: @sternenseemann

Reviewer: ```suggestion
author: @sternenseemann 
co-authors: @Ericson2314
shepherd-team: @L-as @kfearsoff @tomberek @DavHau
shepherd-leader: @tomberek
```
Apologies for the previous suggestion, where I missed out making sternenseemann the author.

Reviewer 4: Wait, I thought @Ericson2314 wrote this?

Reviewer 2: https://github.com/NixOS/rfcs/pull/109#issuecomment-1576424589

related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

Nixpkgs contains non-trivial amounts of generated, rather than hand-written code.
We want to start systematizing to make it easier to maintain.
There is plenty of future work building upon this we could do, but we stop here for now to avoid needing to change any tools (Nix, Hydra, etc.).

# Motivation
[motivation]: #motivation

Nixpkgs, along with every other distro, also faces a looming crisis: new open source software is increasingly not intended to be packaged by distros at all.
Many languages now support very large library ecosystems, with dependencies expressed in a language-specific package manager.

Right now, to deal with these packages, we either convert by hand, or commit lots of generated code into Nixpkgs.
But I don't think either of those options is healthy or sustainable.
The problem with the first is sheer effort; we'll never be able to keep up.
The problem with the second is bloating Nixpkgs but more importantly reproducibility: If someone wants to update that generated code it is unclear how.

> All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.

Reviewer 5: > The problem with the second is bloating Nixpkgs but more importantly reproducibility: If someone wants to update that generated code it is unclear how.

I think this needs an explanation as to why reproducibility affects simplicity of updating the packages.

I don't see a clear relationship between `reproducibility` and the `difficulty updating a package`.
We can make update mechanisms that are very easy to use without guaranteeing reproducibility, or vice versa have reproducible mechanisms which are hard to use. 

For example, an update interface like `nix run .#my-packages.update` would be very simple to use, but its result not necessarily reproducible.

Vice verse, an update mechanism can be perfectly reproducible, but if it lacks a good UI, and instead requires the user to read the source code of nixpkgs to find the right commands or attributes to trigger it, it is difficult to use despite being reproducible.

Currently the RFC seems to focus on reproducibility only. I think the statement about the UX improvement should either be dropped, or an explanation added as to how this is achieved.

> All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.

I think it makes sense to clarify which users are targeted by the RFC. Is it `nixpkgs maintainers` or `downstream users` or both?

What `downstream users` need is a way to create, update and modify package sets to their needs. For them, these interactions need to be simple, more than reproducible. The modifications become manifested in the code anyways and reproducibility is ensured from there on.
Increasing the complexity of tools in order to comply with the RFC can introduce a cost for these users, with no added benefit.

Reviewer 2: > I think this needs an explanation as to why reproducibility affects simplicity of updating the packages.

In the sense that when it is easy to _reproduce_ the state of the generated code as present in the repository, it is easier to also effectuate desired changes. If it is, for example, not possible to update generated code without pulling in all available updates, it may be harder to just change a specific, little thing.

This would be my account of this, but perhaps the paragraph can be improved.

> Is it nixpkgs maintainers

Yes.

The lowest hanging fruit is to systematize our generated code.
We should ensure anyone can update the generated code, which means it should be built in derivations not some ad-hoc way.

> In short, we should apply the same level of rigour that we do for packages themselves to generate code.

Reviewer 6: ```suggestion
In short, we should apply the same level of rigour that we do for packages themselves to generated code.
```

# Detailed design
[design]: #detailed-design

## Nixpkgs

1. Establish the policy that all generated code in nixpkgs must be produced by a derivation.
   The derivation should be built by CI (so exposed as some Nixpkgs in some fashion).

2. Implement script(s) for maintainers which automatically builds these derivations and vendors their results to the appropriate places.
   Running such scripts should be sufficient to regenerated all generated code in Nixpkgs.

   Greenfield tooling should not be merged unless it complies with the policy from day one.
   Existing non-compliant tooling doesn't need to be ripped out of Nixpkgs, but the "grace period" in which is brought to compliance should be bounded.

3. Ensure via CI that the vendored generated code is exactly what running the scripts produce.
   This check should be one of the "channel blocking" CI jobs.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

## Impurities

Many `lang2nix`-type tools have impure steps today.

> Since these tools must be only invoked inside the derivations to generate code, the impure inputs must be gotten via fixed output derivations.

Reviewer 7: I would strongly recommend against Fixed-ouput drvs. they lead to a situation where it’s trivial to run into a reproducibility problem. The prefetching step as mentioned a few lines later is the way to go, since that way only one maintainer generates the file and we are not at the mercy of $version resolver being reproducible.

Reviewer 2: Version solvers etc. should be executed inside the normal derivation. Fixed output derivations would be package repository indices etc.

Author: @Profpatsch I am a bit confused what you mean? I mean fixed output derivations to fetch things, not fixed output derivations that to a lot of arbitrary "vendoring" work.

Reviewer 7: Okay, but then I’d just call them “fixed output fetchers” here, to make the distinction. 

Cause you can also used FODs for doing arbitrary stuff.

Reviewer 4: I have a similar concern (perhaps it is the same concern).  Let me be very concrete and use Rust's `crates.io` as an example.

Rust's `cargo update` does a `git` fetch from the `master` branch of `https://github.com/rust-lang/crates.io-index`.  Right now `buildRustPackage` doesn't pin the revision of this repo -- it still fetches `master` and relies on the fact that the crates.io team never deletes `package:version` entries except when a package is "yanked" due to malware or vulnerabilities.

When crates.io yanks a package, people who use `cache.nixos.org` don't notice, because they hit the nixos cache, which never sees the yank.  People who build without any substituters will get a build failure.  I consider this to be a reproducibility problem.

@Ericson2314 / @sternenseeman: if we were to use `crate2nix`-generated nix code for Rust's `crates.io` (the way we currently use `cabal2nix`-generated code for Haskell's Hackage), this RFC says that code should be generated by a derivation.  Would that derivation fetch `https://github.com/rust-lang/crates.io-index` from a pinned revision or would it use the `master` branch?

Reviewer 4: I have the same (I think) concern.

In order to be totally concrete I will use Rust as an example.

Suppose that we use `crate2nix`-generated nix code for Rust's `crates.io`, the way we currently use `cabal2nix`-generated code for Haskell's Hackage.  Suppose also that we stop using upstream's `Cargo.lock` files in order to be able to apply security updates to deep dependencies (e.g. `ring`, the Rust crypto library), to get better build/disk sharing, and to be able to drop the 74 `Cargo.lock` patches currently in nixpkgs.

In order to generate our own `Cargo.lock` files, the `crate2nix`-running derivation would need to run `cargo update`.  Rust's `cargo update` does a `git` fetch from the `master` branch of `https://github.com/rust-lang/crates.io-index`.  If we take this approach, when crates.io yanks a package, people who use `cache.nixos.org` don't notice, because they hit the nixos cache, which never sees the yank.  People who build without any substituters will get a build failure.

I consider this to be a reproducibility problem.

@Ericson2314 / @sternenseemann : Would that derivation fetch `https://github.com/rust-lang/crates.io-index` from the `master` branch, or would it fetch from a pinned git revision?  In the latter case, is this something we should require as a matter of policy?  If it is required as a matter of policy it should be rephrased in a way that isn't Rust/`crates.io`-specific of course.

Note: an earlier version of this comment contained incorrect statements about `buildRustPackage`.  If you saw that, please ignore it.

Reviewer 8: > we are not at the mercy of $version resolver being reproducible.

Aren't we that anyway no matter how we implement it? Vendored FOD or not? We somehow need to pin dependency versions and that part must be reproducible.

Reviewer 9: I would suggest fixing the wording to say something along the lines of "while making sure that the impure steps are as small as possible", albeit with better wording.

Author: How about this?
```suggestion
Since these tools must be only invoked inside the derivations to generate code, they must operate purely.
Any code that needs to be fetched must be done separately, e.g. with fixed-output derivations.
Any work deciding *what needs to be fetched* in the first place must be done impurely and out of band; this should be as little work as possible.
```

Reviewer 10: There is the new "impure" derivation that might be useful here. If the CI checks for on-disk representation stability, this might help surface reproducibility problems. It also saves from downloading dependencies twice when updating FOD hashes.

Reviewer 2: I don't think that is necessary, in most cases the input addressed approach will be ok: We have fixed-output data files as an input (which is updated via an "impure step" i.e. an update script) and run a solver in a regular derivation (which is hopefully deterministic!).

For taming quicklisp I don't think impure derivations are useful, since they are still sandboxed, aren't they?

Reviewer 11: For Quicklisp we look at the packages themselves to check the dependency information (index contains some claims about dependencies that are usually correct but we have seen some exceptions so we don't trust that blindly)

I don't think Quicklisp has lockfiles: normal mode of operations works within a single globally coordinated (and tested) version set.

Reviewer 5: @Ericson2314 I think your suggested change from above is good and should be applied.

The approach of running a version resolver inside a derivation is not a practical approach for some ecosystems.
The reason usually being that there is no package index readily available to feed into the `solver derivation`.

As mentioned above, some ecosystems expose dependency info only inside package sources, not offering a central index.

We can build our own package index by crawling all packages regularly, but the upstream resolver will most likely be incompatible to our custom package index.
-> we now have to invent our own dependency resolver as well.

I already went through all of that when building [mach-nix](https://github.com/DavHau/mach-nix) which is a nix code generator for python using a custom python package index [pypi-deps-db](https://github.com/DavHau/pypi-deps-db).

My personal experience is that this approach introduces significant maintenance overhead. Not only must the `crawler`/`db`/`resolver` be compatible to the complete history of standardization of the ecosystem but also adopt all upcoming standards to not fall out of sync.

Therefore think it is a good idea for the RFC to allow a minimal amount of `impure operations`. This simply allows to use upstream tooling to lock down the inputs.

This might require changes to those tools to separate the pure work from the impuire gather steps.

> Additionally, as @7c6f434c point out, some upstream tooling thinks it is being pure, but the "lock files" (or similar) pinning mechanism it provides isn't up to the task for Nix's purposes.

Reviewer 12: I can give a specific example of Gradle lockfiles. It specifies the package and its version, but it doesn't specify its hash. There is a separate mechanism for storing dependencies' hashes, but due to the way Gradle works the same version may be fetched from a different repository in a different part of the build graph, which may have a different hash

> Quicklisp, for example, uses a "weird mix of MD5 constraints and SHA1 constraints" that isn't really up to the task.

Reviewer 13: Not clear what issue is being pointed out. Is it about hashes not being cryptographically strong, or that they were forced out of Nix ecosystem, or that it's a mix of the two?

Author: Most generally, any hash we cannot cajoul into a store path or fixed output derivation.

It needs to be a hash algo we support, and (assuming no one else supports NAR) a hash of a flat file (such as a tarball).

Reviewer 9: We can just support the ones we don't right now, but I do think it's mainly cryptographic strength. We don't want to support non-cryptographic hashes, yet we support SHA1 ironically.

> Another example would be using git commit hashes, which, since we don't want to download the whole history, are not good enough on their own.

Reviewer 9: This isn't true. Given a commit, you can fetch the tree corresponding to that commit without fetching any of the parent commits (or their data).

Reviewer 11: And just to make it explicit: indeed the downloaded content can be verified (the previous history is mediated in hashing by just the hash of the parent commit, which is small)

Author: @L-as 
Err i mean the commit hash doesn't work as a fixed output hash. Yes, it does identify what to download, but isn't sufficient to very the downloaded tree in isolation.

@7c6f434c Yes, I mentioned such Merkle inclusion proofs as future work in a comment in the IPFS RFC: https://github.com/NixOS/rfcs/pull/133/files#r957445525. But I think they are well out of scope for this one.

Any idea how I might make this clearer?

Reviewer 11: Removing the part about downloading the whole history and saying that git commit hashes are not supported by Nix would be an improvement. If you want to give a justification why they probably won't be supported really soon, say they are more complicated to verify than normal content hashes. I just agreed that whole-history justification doesn't justify things because it's not completely true; I agree that extra complexity (and git-specificity) justifies lack of Nix support.

> A concrete example of a change that would bring such tooling into compliance is via "prefetching" to build a map of insufficient upstream-tool keys (say a pair of a name and lousy hash) to higher quality hashes for fixed output derivations.

Reviewer 4: ```suggestion
A concrete example of a change that would bring such tooling into compliance is via "map-building" to build a map of insufficient upstream-tool keys (say a pair of a name and lousy hash) to higher quality hashes for fixed output derivations.
```

> The prefetching step would be run impurely but do as little work as possible, and the remaining bulk of the work would be done purely in derivations.

Reviewer 9: :+1:

> The prefetching step would be run impurely but do as little work as possible, and the remaining bulk of the work would be done purely in derivations.

Reviewer 4: ```suggestion
The map-building step would be run impurely but do as little work as possible, and the remaining bulk of the work would be done purely in derivations.
```

> The prefetching step would be run impurely but do as little work as possible, and the remaining bulk of the work would be done purely in derivations.

Reviewer 4: > We'd actually only prefetch the crates.io index

@sternenseemann, fetching the crates.io index is pure (it's an FOD).  So if that's an example of "prefetching", why does the text describe "prefetching" as being "run impurely"?

This terminology is a landmine that we will regret.

Author: I am convinced in getting rid of "prefetch". @sternenseemann among other things the "pre" makes a terminology based on statefulness. This is sort of confusing and generally not how we want people to think about Nix things, so an alternative non-state-oriented term seems like an upgrade to me.

@sternenseemann tells me he must spend a lot of manual time shepherding PRs, because those PR authors are unable to jump through the hoops themselves.

# Drawbacks
[drawbacks]: #drawbacks

This is now a very conservative RFC so I do not think there are any drawbacks as to the goals themselves.

Bringing our tools into compliance with this policy will take effort, and of course that effort could be spent elsewhere, so there is opportunity cost to be aware of.
But given the general level of concern over the sustainability of Nixpkgs, I think the benefits are worth the costs.

# Alternatives
[alternatives]: #alternatives

No good this time, we had other ideas but they are reframed as *possible* future work.
It is unclear which of the alternative "2nd steps" is better, or whether we ought to try to jump ahead straight to the "3rd step".

The plan proposed here is unquestionably the most conservative one, and basically a prerequisite of all the others --- a first step no matter what we plan to do afterwords.

> 

Reviewer 9: Using standard IFD might be worth including.

  E.g. gcrooting this evaluation only dependency could prove tricky without changes to Nix.

- Extra bureaucracy would be involved with updating the generated repository and the reference to it in nixpkgs.
  Additionally, special support in CI would be required for this.

### Alternative 2: Nixpkgs itself becomes a derivation output

This alternative implementation was proposed by @L-as at the meeting.
The idea is that nixpkgs would become a derivation that builds a “regular” nixpkgs source tree by augmenting files available statically with code generation.

The upside of this would be that there would only be one instance of IFD that can ever happen, namely when the source tree is built.
The produced store path then would require no IFD, and it would be obvious what relates to IFD and what doesn't.

In practice, IFD would not be necessary for users of nixpkgs if we can design a mechanism that allows the dynamically produced nixpkgs source tree to be used as a channel.
Then the IFD would only need to be executed when working on nixpkgs.

#### Drawbacks

- This approach creates a bootstrapping problem for the entirety of nixpkgs, not just for the IFD parts.

>   It would be necessary to build the new nixpkgs source tree using an old version of the nixpkgs source tree.

Reviewer 14: This sounds awful, I hate how manual `bootstrap-tools` are, doing it to the whole Nixpkgs feels even worse.

The question, though: Why can't only the generated parts of Nixpkgs be bootstrapped using the rest? That would better than either alternative, IMHO.
In fact, I wanted to do this for `all-packages.nix` in NixOS/nixpkgs#50643 (4 years before #140 and NixOS/nixpkgs#237439).

The ultimate idea (which was abandoned after the initial pushback) there was to eventually generate most of `all-packages.nix` by walking the Nixpkgs tree via a simple shell script.
So, `stdenv` and the knot of packages supporting it would have been left in a tiny hand-written equivalent of `all-packages.nix`, and the rest of `all-packages.nix` would have been auto-generated.

Reviewer 14: > so the bootstrap problem already exists and can not be alleviated easily.

Well, yes, but there would be quite a difference between bootstrapping from an autogenerated `.nix` file that depends on current version of Nixpkgs (and which you can still read and understand, even if it is _technically_ programmatically generated) and bootstrapping from an autogenerated `.nix` file that depends on an outdated version of Nixpkgs that bootstraps from an autogenerated `.nix` file that depends on an even more outdated version of Nixpkgs, which bootstraps....

Reviewer 2: That's why this is listed as an alternative to the actual proposal which does not require that!

## A possible 3rd step: Import from derivation

Even if we store the generated sources outside of tree, we are still doing the tedious work of semi-manually remaining a build cache (this time of Nix code).
Isn't that what Nix itself is for!

"import from derivation" is a technique where Nix code can simply import the result of a build, with no vendoring generated code in-tree or out-of-tree needed.

There are a number of implementation issues with it, however, that means we can't simply enable it on `hydra.nixos.org` today.
We have some "low tech" mitigations that were the original body of this RFC,
but they still require changing tools (Hydra), which adds latency and risk to the project.

## Getting upstream tools to agree on how to pin source code

A source of frustration outlined in the [Impurities](#impurities) section is when upstream tools think they are pinning exactly dependencies down, but nonetheless do so in a way that isn't good enough for our purposes.
A long standing goal of mine is to try to communicate these concerns back upstream, and nudge everyone agreeing on a common definition of what a pinned deps looks like.

> I think policies such as this RFC proposes will allow us to get our `lang2nix` infrastructure in a more state not only more legible to ourselves (Nix users and contributors) but also upstream developers who won't want to spend too long investigating what exactly our requirements are.

Reviewer 6: ```suggestion
I think policies such as this RFC proposes will allow us to get our `lang2nix` infrastructure in a state not only more legible to ourselves (Nix users and contributors) but also to upstream developers who won't want to spend too long investigating what exactly our requirements are.
```
